### PR TITLE
apidoc: Move apidoc to ext/apidoc

### PR DESCRIPTION
**Pull request by stephenfin**
The 'sphinx-apidoc' tool is no longer the only kid on the block when it
comes to automatic documentation of Python projects, with the likes of
'sphinx-autoapi' in development [1] and is not really maintained. Given
the fact that the tool is tangential to Sphinx's main goal, there isn't
really anything that warrants its continue existence as a core module.

Signify this by moving the feature to 'ext'. This allows the feature to
continue to exist in the package, but indicates that stability might be
worse than one would expect from the core library.

To avoid breaking packages that are using this feature directly, such as
pbr [3], aliases for the old 'main' method are included. This is based
on what Django does and, like Django, will allow us to safely remove the
old modules in Sphinx 2.0.

[1] https://github.com/rtfd/sphinx-autoapi
[2] https://github.com/sphinx-doc/sphinx/pull/3485#issuecomment-288081223
[3] https://github.com/django/django/blob/1.11/django/test/runner.py#L688-L695

### CODE REVIEW

File: sphinx/ext/apidoc.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
+# -*- coding: utf-8 -*-
+"""
+    sphinx.apidoc
```
Comment (tk0miya) [inline sphinx/ext/apidoc.py]: This should be `sphinx.ext.apidoc`.

File: sphinx/ext/apidoc.py [hunk 2]
```diff
@@ -0,0 +1,439 @@
+# -*- coding: utf-8 -*-
```
Comment (tk0miya) [inline sphinx/ext/apidoc.py:1]: It seems this is a new file.
Please rename from `sphinx/apidoc.py` and create `sphinx/apidoc.py` as a new file to keep commit history,

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits.

**Comment by tk0miya** [comment]
It seems travis raises errors. please check them.

@shimizukawa Do you have any comments?
I wonder which is better `sphinx.ext.apidoc` or `sphinx.ext.autodoc.apidoc`. what do you think?

**Comment by shimizukawa** [comment]
I have no objection. 
And +1 to `sphinx.ext.apidoc`. `sphinx.ext.autodoc.apidoc` is too deep to me.

**Review by shimizukawa — COMMENTED** [review]
There is no more comment than @tk0miya's point.

**Comment by tk0miya** [comment]
Thanks, either is okay. I think `sphinx.ext.autodoc.apidoc` is better if apidoc is a part of autodoc. But I don't know so much about apidoc. so it is not strong opinion.
Let's go with `sphinx.ext.apidoc`.

File: sphinx/ext/apidoc.py (hunk 2 above)
Comment (stephenfin) [reply sphinx/ext/apidoc.py:1]: I'm not able to do anything about this. Git does file rename detection automatically (`git [mv]` might well be an alias for `git cp [src] [dst]` and `git rm [src]`), and its heuristics seem to be failing us here.

File: sphinx/ext/apidoc.py (hunk 1 above)
Comment (stephenfin) [reply sphinx/ext/apidoc.py]: Done

**Comment by stephenfin** [comment]
@tk0miya OK, I've pushed the reworked version up and the build looks to be good now (it was a pep8 issue). Think that's all that's needed now?

**Comment by stephenfin** [comment]
> Thanks, either is okay. I think sphinx.ext.autodoc.apidoc is better if apidoc is a part of autodoc. But I don't know so much about apidoc. so it is not strong opinion.

It is and it isn't :smile: The `sphinx-apidoc` tool generates documents that contain `sphinx.ext.autodoc` directives. However, it's not really part of `sphinx.ext.autodoc`, per se. If we were to move this, we should probably move `sphinx.ext.autosummary` too, as that also generates stub documents with `sphinx.ext.autodoc` directives in them, but I wouldn't suggest doing that.

In the long term, I'd really like to see all three packages moved out of the Sphinx core and into their own package(s) in the [sphinx-doc namespace](https://github.com/sphinx-doc), for all the reasons given in the commit message. I think a separate package would be a better way of indicating that this is very much an add-on to Sphinx, rather than a key, core piece of functionality. However, I'm not in the position to make this call myself nor would I be confident enough with the source (yet!) to maintain this new package myself. I'll wait til I've more work done with the Sphinx codebase before committing myself here, heh.

**Comment by stephenfin** [comment]
Could I get some eyes on this again? I think I've resolved everything that needs resolving?

**Review by tk0miya — APPROVED** [review]
Sorry for late response. I'm in busy last 2-3 months. But I'm back now :-)

**Comment by tk0miya** [comment]
Merged. Thank you for your good work!

**Pull request merged**